### ISSUE TITLE
Cannot properly mock injected class from constructor

### ISSUE THREAD

**Issue by betaWeb** [laravel/framework#52401]
### Laravel Version

9.52.16

### PHP Version

8.1.22

### Description

Hello 👋

I wan't to mock a simple cache repository class, that is injected in another class constructor (and used in several methods of this class) like so :
```php
class FooService // dummy class name for the example
{
    public function __construct(private readonly CacheRepository $cacheRepository)
    {
    }
    
    public function doStuffWhichCallsInjectedCacheRepository(): void
    {
      // ...
      $value = $this->cacheRepository->getCacheValue(...);
      // ...
    }
}
```

In my test, I wan't to mock it :
```php
$ids = [1,2,3,4,5]; // dummy data for the example
$this->partialMock(CacheRepository::class)
    ->shouldReceive('getCacheValue')
    ->times(5)
    ->andReturn(...$ids);

app(FooService::class)->doStuffWhichCallsInjectedCacheRepository();
```

So the class (`FooService` in this example) in which the cache repository is injected is called, has his method correctly called and stuff, but the mock does not work properly.

**However**, if I edit my class `FooService` with an on-the-fly `CacheRepository` class resolution from IOC, like so, my tests works properly :
```php
class FooService // dummy class name for the example
{    
    public function doStuffWhichCallsInjectedCacheRepository(): void
    {
      $cacheRepository = app(CacheRepository::class);
      
      // ...
      
      $value = $cacheRepository->getCacheValue(...);
      
      // ...
    }
}
```

In conclusion it seems that dependencies injected in constructor cannot be mocked properly, but for on-the-fly resolved dependencies it's ok.

If anybody has an explanation or a workaround to allow both dependencies injected in a constructor AND mock works together.. thanks !

### Steps To Reproduce

/
